File: Solutions/117. Populating Next Right Pointers in Each Node II/117. Populating Next Right Pointers in Each Node II.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        queue<Node*> q;
        if(root){
            q.push(root);
        }
        while(q.size()){
            int sz=q.size();
            Node* prev=nullptr;
            for(int i=0;i<sz;++i){
                auto cur=q.front();q.pop();
                cur->next=prev;
                prev=cur;
                if(cur->right){
                    q.push(cur->right);
                }
                if(cur->left){
                    q.push(cur->left);
                }
            }
        }
        return root;
    }
};

File: Solutions/117. Populating Next Right Pointers in Each Node II/117. Populating Next Right Pointers in Each Node II.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
    var q []*Node
    if root!=nil{
        q=append(q,root)
    }
    for len(q)>0{
        sz:=len(q)
        var prev *Node
        prev=nil
        for i:=0;i<sz;i+=1{
            cur:=q[0];q=q[1:]
            cur.Next=prev
            prev=cur
            if cur.Right!=nil{
                q=append(q,cur.Right)
            }
            if cur.Left!=nil{
                q=append(q,cur.Left)
            }
        }
    }
    return root
}